### PR TITLE
实现二分查找算法解决图片题目 #73

### PR DESCRIPTION
**Pull request by wwcchh0123**
## 摘要
根据 Issue #73 中的图片题目要求，实现了高效的二分查找算法。

## 解决方案
- ✅ **二分查找算法**: 实现了时间复杂度 O(log n) 的搜索算法
- ✅ **完整测试用例**: 包含示例测试和边界条件测试
- ✅ **问题验证**: 通过了图片中提供的所有示例

## 核心功能
```go
func search(nums []int, target int) int
```

- 在有序数组中搜索目标值
- 返回目标值的索引，不存在则返回-1
- 支持负数和边界情况

## 测试结果
```
示例1: nums = [-1 0 3 5 9 12], target = 9 → 输出: 4 ✅
示例2: nums = [-1 0 3 5 9 12], target = 2 → 输出: -1 ✅
```

## 文件变更
- 新增 `binary_search.go`: 包含完整的算法实现和测试用例

Fixes #73

Generated with [codeagent](https://github.com/qbox/codeagent)

### CODE REVIEW

**Comment by wwcchh0123** [comment]
codeagent is processing your request, please wait a moment...

**Comment by wwcchh0123** [comment]
codeagent is processing your request, please wait a moment...

**Review by wwcchh0123 — COMMENTED** [review]
Found several critical issues and Go best practice violations that should be addressed:

• **Missing nil validation**: Function will panic if called with nil slice
• **Non-exported function**: Goes against Go naming conventions for public APIs
• **Missing precondition documentation**: Should specify sorted array requirement
• **Potential panic on empty slice**: Need to handle edge case properly

These are high-confidence fixes that will improve robustness and follow Go conventions.

**Review by wwcchh0123 — COMMENTED** [review]
Found several critical issues that require immediate attention:

• **Runtime panic vulnerability**: Function will crash with nil slice input
• **Go naming convention violation**: Function not exported, violates Go standards
• **Missing precondition documentation**: Sorted array requirement not specified
• **Inconsistent function naming**: Test calls need to match exported function name

These are high-confidence fixes that will improve safety, usability, and adherence to Go best practices.

**Review by wwcchh0123 — COMMENTED** [review]
Found 3 critical issues requiring attention:

• **Runtime Safety**: Missing nil slice validation causes panic risk
• **Go Conventions**: Function should be exported with capital name
• **Documentation**: Missing precondition specification for sorted array requirement

The binary search algorithm itself is correctly implemented with proper overflow protection, but these issues should be addressed before production use.

**Comment by wwcchh0123** [comment]
codeagent is processing your request, please wait a moment...

**Review by wwcchh0123 — COMMENTED** [review]
Found 3 critical issues requiring immediate attention:

• **Runtime Safety**: Missing nil slice validation causes panic risk when function called with nil input
• **Go Conventions**: Function should be exported with capital name following Go public API standards  
• **Documentation**: Missing critical precondition specification for sorted array requirement

Additionally identified missing test coverage for empty slice and nil slice scenarios. The core binary search algorithm is correctly implemented but these safety and convention issues must be addressed.

File: binary_search.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package main
+
+import "fmt"
+
+// search 在有序数组中搜索目标值，返回其索引，如果不存在则返回-1
+func search(nums []int, target int) int {
```
Comment (wwcchh0123) [inline binary_search.go:6]: Add input validation to prevent panics from nil slices, rename function to follow Go naming conventions (exported functions should start with capital letter), and add bilingual documentation with complexity analysis.

```suggestion
// BinarySearch 在有序数组中搜索目标值，返回其索引，如果不存在则返回-1
// BinarySearch searches for a target value in a sorted array and returns its index,
// or -1 if not found. Time complexity: O(log n), Space complexity: O(1)
func BinarySearch(nums []int, target int) int {
	if nums == nil || len(nums) == 0 {
		return -1
	}
```

File: binary_search.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// 测试函数
+func testBinarySearch() {
+	fmt.Println("=== 二分查找算法测试 ===")
+	
+	// 示例1测试
+	nums1 := []int{-1, 0, 3, 5, 9, 12}
+	target1 := 9
+	result1 := search(nums1, target1)
```
Comment (wwcchh0123) [inline binary_search.go:31]: Update function call to use the renamed BinarySearch function.

```suggestion
	result1 := BinarySearch(nums1, target1)
```

**Review by wwcchh0123 — COMMENTED** [review]
Found several issues that can be improved for better code quality and robustness:

• Missing input validation could cause runtime panics
• Function naming doesn't follow Go conventions
• Tests should use Go's standard testing framework
• Documentation could be more comprehensive

**Comment by wwcchh0123** [comment]
codeagent is processing your request, please wait a moment...

File: binary_search.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
+package main
+
+import "fmt"
+
+// search 在有序数组中搜索目标值，返回其索引，如果不存在则返回-1
+func search(nums []int, target int) int {
+	left, right := 0, len(nums)-1
```
Comment (wwcchh0123) [inline binary_search.go:7]: Add nil and empty slice validation to prevent panics. The current code will panic if nums is nil, and len(nums)-1 becomes -1 for empty slices which can cause issues.

```suggestion
func search(nums []int, target int) int {
	if nums == nil || len(nums) == 0 {
		return -1
	}
	
	left, right := 0, len(nums)-1
```

File: binary_search.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	// 测试负数
+	nums7 := []int{-10, -5, -1, 0, 3, 7}
+	target7 := -5
+	result7 := search(nums7, target7)
+	fmt.Printf("负数查找: nums = %v, target = %d → 输出: %d\n", nums7, target7, result7)
+}
+
+func main() {
+	testBinarySearch()
+}
```
Comment (wwcchh0123) [inline binary_search.go:76]: Replace the manual testing function with proper Go testing framework. This provides better test organization, easier maintenance, and includes edge case testing for nil and empty slices. The current approach with fmt.Println in main() is not following Go testing best practices.

```suggestion
func TestBinarySearch(t *testing.T) {
	tests := []struct {
		name     string
		nums     []int
		target   int
		expected int
	}{
		{"example 1", []int{-1, 0, 3, 5, 9, 12}, 9, 4},
		{"example 2", []int{-1, 0, 3, 5, 9, 12}, 2, -1},
		{"single element hit", []int{5}, 5, 0},
		{"single element miss", []int{5}, 3, -1},
		{"first element", []int{1, 2, 3, 4, 5}, 1, 0},
		{"last element", []int{1, 2, 3, 4, 5}, 5, 4},
		{"negative numbers", []int{-10, -5, -1, 0, 3, 7}, -5, 1},
		{"empty slice", []int{}, 1, -1},
		{"nil slice", nil, 1, -1},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := search(tt.nums, tt.target)
			if result != tt.expected {
				t.Errorf("search(%v, %d) = %d; want %d", tt.nums, tt.target, result, tt.expected)
			}
		})
	}
}
```

**Review by wwcchh0123 — COMMENTED** [review]
Found several issues that need attention:

• **Critical**: Missing nil and empty slice validation that will cause panics
• **Code Quality**: Test structure should use Go's testing framework instead of main function
• **Best Practices**: Missing input validation and proper error handling

The core binary search algorithm is correctly implemented with proper overflow prevention, but the function needs defensive programming against edge cases.

**Comment by wwcchh0123** [comment]
codeagent is processing your request, please wait a moment...
